hw26/Vector2d.cpp

```diff
@@ -0,0 +1,135 @@
+#include "Vector2d.h"
+#include <iostream>
+
+int Vector2d::activeInstancesCount = 0;
+
+Vector2d::Vector2d() : m_x(0.0f), m_y(0.0f) {
+    activeInstancesCount++;
+}
+
+
+Vector2d::Vector2d(float x, float y) : m_x(x), m_y(y) {
+    activeInstancesCount++;
+}
+
+// new
+Vector2d::Vector2d(float x0, float y0, float x1, float y1) : m_x(x1 - x0), m_y(y1 - y0) {
+    activeInstancesCount++;
+}
+// end new
+
+Vector2d::~Vector2d() {
+    activeInstancesCount--;
+}
+
+Vector2d& Vector2d::operator=(const Vector2d& other) {
+    if (this != &other) {
+        m_x = other.m_x;
+        m_y = other.m_y;
+    }
+    return *this;
+}
+
+Vector2d Vector2d::operator+(const Vector2d& other) const {
+    return Vector2d(m_x + other.m_x, m_y + other.m_y);
+}
+
+Vector2d Vector2d::operator-(const Vector2d& secondVector) const {
+    return Vector2d(m_x - secondVector.m_x, m_y - secondVector.m_y);
+}
+
+float Vector2d::operator()() const {
+    return std::sqrt(m_x * m_x + m_y * m_y);
+}
+
+float& Vector2d::operator[](int i) {
+    if (i == 0) {
+        return m_x;
+    }
+    if (i == 1) {
+        return m_y;
+    }
+    throw std::out_of_range("Index out of range.");
+}
+
+Vector2d operator+(const Vector2d& leftVector, const Vector2d& rightVector) {
+    return Vector2d(leftVector.m_x + rightVector.m_x, leftVector.m_y + rightVector.m_y);
+}
+
+Vector2d operator-(const Vector2d& leftVector, const Vector2d& rightVector) {
+    return Vector2d(leftVector.m_x - rightVector.m_x, leftVector.m_y - rightVector.m_y);
+}
+
+std::ostream& operator<<(std::ostream& os, const Vector2d& vector) {
+    os << "{" << vector.m_x << "; " << vector.m_y << "}";
+    return os;
+}
+
+std::istream& operator>>(std::istream& is, Vector2d& vec) {
+    is >> vec.m_x >> vec.m_y;
+    return is;
+}
+
+float Vector2d::getX() const {
+    return m_x;
+}
+
+float Vector2d::getY() const {
+    return m_y;
+}
+
+void Vector2d::setX(float x) {
+    m_x = x;
+}
+
+void Vector2d::setY(float y) {
+    m_y = y;
+}
+
+void Vector2d::printInfo() const {
+    std::cout << "(" << m_x << ", " << m_y << ")" << std::endl;
+}
+
+int Vector2d::getActiveInstancesCount() {
+    return activeInstancesCount;
+}
+
+// new
+float Vector2d::dotProduct(const Vector2d& other) const {
+    return m_x * other.m_x + m_y * other.m_y;
+}
+
+Vector2d Vector2d::negate() const {
+    return Vector2d(-m_x, -m_y);
+}
+
+VectorRelativeState Vector2d::getRelativeState(const Vector2d& other) const {
+    if (m_x == other.m_x && m_y == other.m_y) {
+        return VectorRelativeState::Identical;
+    }
+
+    float scalarProduct = dotProduct(other);
+    float magnitude1 = (*this)();
+    float magnitude2 = other();
+
+    float cos = scalarProduct / (magnitude1 * magnitude2);
+
+    if (cos == 1.0f) {
```

Comment:
comparing float directly in general is not that simple. You have to compare it up to some presicion.
see https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon